generate-spec

```diff
@@ -680,18 +714,91 @@ if (count($schemas) == 0 && count($routes) == 0) {
 	Logger::error("app", "No spec generated");
 }
 
-if (count($openapi["paths"]) == 0) {
-	$openapi["paths"] = new stdClass();
-}
-
 ksort($schemas);
-$openapi["components"]["schemas"] = count($schemas) == 0 ? new stdClass() : $schemas;
 
 if ($useTags) {
 	$openapi["tags"] = $tags;
 }
 
-file_put_contents($out, json_encode($openapi, Helpers::jsonFlags()));
+foreach ($scopePaths as $scope => $paths) {
+	$openapiScope = $openapi;
+
+	if (count($paths) == 0) {
+		$paths = new stdClass();
+	}
+
+	$scopeSuffix = $scope === 'default' ? '' : '-' . $scope;
+	$openapiScope['info']['title'] .= $scopeSuffix;
+	$openapiScope['paths'] = $paths;
+
+	$usedSchemas = [];
+	foreach ($paths as $url => $urlRoutes) {
+		foreach ($urlRoutes as $httpMethod => $routeData) {
+			foreach ($routeData['responses'] as $statusCode => $responseData) {
+				if (isset($responseData['content']['application/json'])) {
+					$usedSchemas = array_merge($usedSchemas, Helpers::collectUsedRefs($responseData['content']['application/json']['schema']));
+				} elseif (isset($responseData['content']['*/*']['schema']['type'], $responseData['content']['*/*']['schema']['format'])
+					&& $responseData['content']['*/*']['schema']['type'] === 'string'
+					&& $responseData['content']['*/*']['schema']['format'] === 'binary') {
+					Logger::info("app", "Binary response from '$httpMethod $url' - Skipping schema reading");
+				} else {
+					Logger::warning("app", "Could not read used schemas for response to '$httpMethod $url' with status code $statusCode");
+				}
+			}
+		}
+	}
+
+	$scopedSchemas = [];
+	while ($usedSchema = array_shift($usedSchemas)) {
+		if (!str_starts_with($usedSchema, '#/components/schemas/')) {
+			continue;
+		}
+
+		$schemaName = substr($usedSchema, strlen('#/components/schemas/'));
+
+		if (!isset($schemas[$schemaName])) {
+			Logger::error("app", "Schema $schemaName used by scope $scope is not defined");
+		}
+
+		// Queue potential sub-refs for exporting as well
+		foreach (['allOf', 'oneOf', 'anyOf', 'properties'] as $group) {
```

Comment:
Still not enough, for example `additionalProperties` can also contain a schema. I would say to be save it is best to really traverse all the schemas otherwise some might be missed.

Reply:
looping over all does not work, because `items` e.g. needs to be treated differently then those others. But I checked all from https://github.com/nextcloud/openapi-extractor/blob/a07d21cbc1f731ed4c1359a64e5c79db5203383d/src/OpenApiType.php#L73-L94 now.